midend/parserUnroll.cpp

```diff
@@ -367,6 +366,38 @@ class ParserSymbolicInterpreter {
             auto mc = sord->to<IR::MethodCallStatement>();
             auto e = ev.evaluate(mc->methodCall, false);
             success = reportIfError(state, e);
+        } else if (auto bs = sord->to<IR::BlockStatement>()) {
+            IR::IndexedVector<IR::StatOrDecl> newComponents;
+            for (auto* component : bs->components) {
+                auto newComponent = executeStatement(state, component, valueMap);
+                if (!newComponent)
+                    success = false;
+                else
+                    newComponents.push_back(newComponent);    
+            }
+            sord = new IR::BlockStatement(newComponents);
+        } else if (sord->is<IR::IfStatement>()) {
```

Comment:
is this needed? 
there are passes which eliminate if statements from parsers.

Reply:
This option is called in issue1897-bmv2.p4 and is associated with header_union

Reply:
If this fixes an issue you will have to add test cases to show that.

Reply:
Ok, I`ll do it

Reply:
Where is the test case that you have added?

Reply:
Is it worth adding another test if issue1897-bmv2.p4 triggers the actions performed in our changes?

Reply:
Yes, especially if that particular test isolates the case you want to implement. issue1897 is intended to cover other features.

Reply:
Ok, I`ll do it

Reply:
Can you show a test example which after the midend transformations has an if statement in the parser?

Reply:
parser-unroll-test6.p4 or issue1897-bmv2.p4

Reply:
Oh I meant, what does the code look like after the transformations. So dumping with `p4test --top4 "End" --dump dmp` and then checking what the output before `parserUnroll` looks like. If there is an if statement left, it looks like a bug in the parser pass that removes parser ifs?

Reply:
Right before ParserUnroll there remains an if, just checked. So there is no point in the check I added for IfStatement, because the problem is elsewhere?

Reply:
Can you paste the code? 
@mbudiu-vmw RemoveParserIfs is a frontend pass, do you think this is an issue?

Reply:
What code exactly?

Reply:
Yes, the section that still had the if statement.

Reply:
```
parser ProtParser(packet_in packet, out headers hdr, inout metadata meta, inout standard_metadata_t standard_metadata) {
    @name("ProtParser.addrType_0") bit<8> addrType_0;
    @name("ProtParser.addr_0") addr_t addr_0;
    state start {
        packet.extract<addr_type_t>(hdr.addr_type);
        addrType_0 = hdr.addr_type.dstType;
        addr_0.ipv4.setInvalid();
        addr_0.ipv6.setInvalid();
        transition ProtAddrParser_start;
    }
    state ProtAddrParser_start {
        transition select(addrType_0) {
            8w0x1: ProtAddrParser_ipv4;
            8w0x2: ProtAddrParser_ipv6;
        }
    }
    state ProtAddrParser_ipv4 {
        packet.extract<addr_ipv4_t>(addr_0.ipv4);
        transition start_0;
    }
    state ProtAddrParser_ipv6 {
        packet.extract<addr_ipv6_t>(addr_0.ipv6);
        transition start_0;
    }
    state start_0 {
        if (hdr.addr_dst.ipv4.isValid()) {
            addr_0.ipv4.setValid();
            hdr.addr_dst.ipv4.addr = addr_0.ipv4.addr;
        } else {
            addr_0.ipv4.setInvalid();
        }
        if (hdr.addr_dst.ipv6.isValid()) {
            addr_0.ipv6.setValid();
            hdr.addr_dst.ipv6.addr = addr_0.ipv6.addr;
        } else {
            addr_0.ipv6.setInvalid();
        }
        addrType_0 = hdr.addr_type.srcType;
        addr_0.ipv4.setInvalid();
        addr_0.ipv6.setInvalid();
        transition ProtAddrParser_start_0;
    }
    state ProtAddrParser_start_0 {
        transition select(addrType_0) {
            8w0x1: ProtAddrParser_ipv4_0;
            8w0x2: ProtAddrParser_ipv6_0;
        }
    }
    state ProtAddrParser_ipv4_0 {
        packet.extract<addr_ipv4_t>(addr_0.ipv4);
        transition start_1;
    }
    state ProtAddrParser_ipv6_0 {
        packet.extract<addr_ipv6_t>(addr_0.ipv6);
        transition start_1;
    }
    state start_1 {
        if (hdr.addr_src.ipv4.isValid()) {
            addr_0.ipv4.setValid();
            hdr.addr_src.ipv4.addr = addr_0.ipv4.addr;
        } else {
            addr_0.ipv4.setInvalid();
        }
        if (hdr.addr_src.ipv6.isValid()) {
            addr_0.ipv6.setValid();
            hdr.addr_src.ipv6.addr = addr_0.ipv6.addr;
        } else {
            addr_0.ipv6.setInvalid();
        }
        transition accept;
    }
}
```

Reply:
> Can you paste the code?
> @mbudiu-vmw RemoveParserIfs is a frontend pass, do you think this is an issue?

There is at least one other pass which introduces ifs, but they should be immediately eliminated. That's the approach i would take.